### PR TITLE
fix(PeriodDialogInputPeriod): Crash in DataInputPeriod. It is not working

### PR DESCRIPTION
**Pull request by MarinaCape**
https://jira.dhis2.org/browse/ANDROAPP-2613

### CODE REVIEW

File: app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt [hunk 1]
```diff
@@ -70,5 +79,38 @@ private class PeriodAdapter(
...
+                }
+                else if (currentDate.before(inputPeriods[inputPeriods.size - 1].initialPeriodDate()) || currentDate.before(
+                        inputPeriodModel.initialPeriodDate()
+                    )
+                )
+                    break
+                else {
+                    currentDate = DateUtils.getInstance().getNextPeriod(periodType, currentDate, -1)
+                }
+            } while (!isAllowed)
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt]: I think is eassier to follow this code if you change the "do while" to a "for" loop :) 

Please when adding so much logic its better to be in classes that this can be testest, this could have different corner cases that could be identified during unit tests. We need to avoid adding logic to android components :sob:

File: app/src/main/java/org/dhis2/utils/customviews/PeriodDialogInputPeriod.java [hunk 2]
```diff
@@ -65,23 +65,31 @@ else if (getCurrentDate().before(inputPeriod.get(inputPeriod.size() - 1).initial
...
             if (isAllowed) break;
         }
 
         if (!isAllowed && !isEmpty) {
             binding.noPeriods.setText(getString(R.string.there_is_no_available_period));
-        } else {
+        }else{
+            boolean withInputPeriod = false;
+
+            if(isAllowed && !isEmpty){
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/PeriodDialogInputPeriod.java]: what is actually "allowed"?

File: app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt [hunk 3]
```diff
@@ -70,5 +79,38 @@ private class PeriodAdapter(
...
+                    ))
+                ) {
+                    isAllowed = true
+                    datePeriods.add(currentDate)
+                    lastDate = currentDate
+                }
+                else if (currentDate.before(inputPeriods[inputPeriods.size - 1].initialPeriodDate()) || currentDate.before(
+                        inputPeriodModel.initialPeriodDate()
+                    )
+                )
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt]: not sure if this pass the klint I believe it will tell you need to add "{"

File: app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt [hunk 4]
```diff
@@ -70,5 +79,38 @@ private class PeriodAdapter(
...
+                if ((currentDate.after(inputPeriodModel.initialPeriodDate()) || currentDate == inputPeriodModel.initialPeriodDate()) && currentDate.before(
+                        inputPeriodModel.endPeriodDate()
+                    )
+                    && (inputPeriodModel.openingDate() == null || inputPeriodModel.openingDate() != null && DateUtils.getInstance().today.after(
+                        inputPeriodModel.openingDate()
+                    )
+                            || DateUtils.getInstance().today == inputPeriodModel.openingDate())
+                    && (inputPeriodModel.closingDate() == null || inputPeriodModel.closingDate() != null && DateUtils.getInstance().today.before(
+                        inputPeriodModel.closingDate()
+                    ))
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt]: I think readability improves a bit if you do this :  

fun isTodayBetweenInitialAndEndPeriod() : Boolean{
return ...
}
fun isTodayBetweenOpenAndCloseDate() :  Boolean{
return ...
}

And then

if (isTodayBetweenInitialAndEndPeriod() &&  isTodayBetweenOpenAndCloseDate()) {
                    isAllowed = true
                    datePeriods.add(currentDate)
                    lastDate = currentDate
}

Comment (MarinaCape) [reply app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt]: You're rigth!
